File: backend/src/routes/itineraries.js
import { Router } from 'express';
import { prisma } from '../db/prisma.js';
import { requireAuth } from '../middleware/auth.js';
import { asyncHandler } from '../middleware/asyncHandler.js';
import { sameRequirements } from '../services/requirementsSnapshot.js';
import { itineraryChoiceSchema, validationError } from '../validation.js';

const router = Router();
router.use(requireAuth);

/**
 * Salva come "draft" l'opzione (primary|alternative) proposta in una conversazione,
 * così da poterla referenziare in fase di prenotazione con un id stabile.
 */
router.post('/', asyncHandler(async (req, res) => {
  const parsed = itineraryChoiceSchema.safeParse(req.body);
  if (!parsed.success) return res.status(400).json(validationError('Dati itinerario non validi', parsed.error));
  const { conversationId, choice } = parsed.data; // choice: "primary" | "alternative"
  const conversation = await prisma.conversation.findUnique({ where: { id: conversationId } });
  if (!conversation || conversation.userId !== req.userId) {
    return res.status(404).json({ error: 'Conversazione non trovata' });
  }
  const result = conversation.state.lastResult;
  if (!sameRequirements(result?.requirementsSnapshot, conversation.state.requirements)) {
    return res.status(409).json({ error: 'La proposta non corrisponde ai requisiti attuali. Genera un nuovo itinerario.' });
  }
  const option = result?.[choice];
  if (!option || option.status !== 'ok') {
    return res.status(400).json({ error: 'Nessuna opzione valida da salvare per questa scelta' });
  }

  const itinerary = await prisma.itinerary.create({
    data: {
      userId: req.userId,
      conversationId,
      status: 'draft',
      totalCost: option.totalCost,
      flightCost: option.breakdown.flightCost,
      hotelCost: option.breakdown.hotelCost,
      activityCost: option.breakdown.activityCost,
      compromises: option.compromises || [],
      details: {
        flights: option.flights,
        destinationMedia: option.destinationMedia || null,
        hotel: { id: option.hotel.hotel.id, name: option.hotel.hotel.name, nights: option.hotel.nights },
        activities: option.activities,
        participants: result.requirementsSnapshot.participants,
        requirementsSnapshot: result.requirementsSnapshot,
      },
    },
  });

  res.status(201).json(itinerary);
}));

router.get('/:id', asyncHandler(async (req, res) => {
  const itinerary = await prisma.itinerary.findUnique({ where: { id: req.params.id } });
  if (!itinerary || itinerary.userId !== req.userId) {
    return res.status(404).json({ error: 'Itinerario non trovato' });
  }
  res.json(itinerary);
}));

export default router;

File: backend-python/app/itineraries.py
from typing import Annotated
from uuid import UUID, uuid4

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from psycopg.types.json import Jsonb

from .database import connect
from .security import current_user_id


router = APIRouter(prefix="/itineraries", tags=["itineraries"])
UserId = Annotated[str, Depends(current_user_id)]


class ItineraryChoice(BaseModel):
    conversationId: UUID
    choice: str = Field(pattern="^(primary|alternative)$")


def serialize(row: tuple) -> dict:
    return {
        "id": row[0], "userId": row[1], "conversationId": row[2], "status": row[3],
        "details": row[4], "totalCost": row[5], "flightCost": row[6],
        "hotelCost": row[7], "activityCost": row[8], "compromises": row[9],
        "createdAt": row[10],
    }


@router.post("", status_code=status.HTTP_201_CREATED)
def save_itinerary(payload: ItineraryChoice, user_id: UserId) -> dict:
    with connect() as connection:
        conversation = connection.execute(
            'SELECT "state" FROM "Conversation" WHERE "id" = %s AND "userId" = %s',
            (str(payload.conversationId), user_id),
        ).fetchone()
        if not conversation:
            raise HTTPException(status_code=404, detail="Conversazione non trovata")
        state = conversation[0] or {}
        result = state.get("lastResult", {}) if isinstance(state, dict) else {}
        option = result.get(payload.choice) if isinstance(result, dict) else None
        if not option or option.get("status") != "ok":
            raise HTTPException(status_code=400, detail="Nessuna opzione valida da salvare per questa scelta")
        breakdown = option.get("breakdown", {})
        requirements = result.get("requirementsSnapshot", {})
        hotel_option = option.get("hotel") or {}
        hotel_info = hotel_option.get("hotel") or hotel_option
        details = {
            "flights": option.get("flights", []),
            "destinationMedia": option.get("destinationMedia"),
            "hotel": {
                "id": hotel_info.get("id"),
                "name": hotel_info.get("name"),
                "nights": hotel_option.get("nights", []),
                "prices": hotel_option.get("prices", {}),
                "pricePerNight": hotel_option.get("pricePerNight"),
            },
            "activities": option.get("activities", []),
            "participants": requirements.get("participants"),
            "requirementsSnapshot": requirements,
        }
        itinerary_id = str(uuid4())
        row = connection.execute(
            'INSERT INTO "Itinerary" ("id", "userId", "conversationId", "status", "details", '
            '"totalCost", "flightCost", "hotelCost", "activityCost", "compromises", "createdAt") '
            'VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, NOW()) '
            'RETURNING "id", "userId", "conversationId", "status", "details", "totalCost", '
            '"flightCost", "hotelCost", "activityCost", "compromises", "createdAt"',
            (itinerary_id, user_id, str(payload.conversationId), "draft", Jsonb(details),
             option.get("totalCost", 0), breakdown.get("flightCost", 0),
             breakdown.get("hotelCost", 0), breakdown.get("activityCost", 0),
             Jsonb(option.get("compromises", []))),
        ).fetchone()
        connection.commit()
    return serialize(row)


@router.get("/{itinerary_id}")
def get_itinerary(itinerary_id: UUID, user_id: UserId) -> dict:
    with connect() as connection:
        row = connection.execute(
            'SELECT "id", "userId", "conversationId", "status", "details", "totalCost", '
            '"flightCost", "hotelCost", "activityCost", "compromises", "createdAt" '
            'FROM "Itinerary" WHERE "id" = %s AND "userId" = %s',
            (str(itinerary_id), user_id),
        ).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Itinerario non trovato")
    return serialize(row)
